src/Appwrite/Platform/Workers/Messaging.php

```diff
@@ -82,85 +85,112 @@ public function action(Message $message, Log $log, Database $dbForProject): void
+        $topicIds = $message->getAttribute('topics', []);
+        $targetIds = $message->getAttribute('targets', []);
+        $userIds = $message->getAttribute('users', []);
 
         /**
-         * @var Document[] $recipients
+         * @var array<Document> $recipients
          */
         $recipients = [];
 
-        if (\count($topicsId) > 0) {
-            $topics = $dbForProject->find('topics', [Query::equal('$id', $topicsId)]);
+        if (\count($topicIds) > 0) {
+            $topics = $dbForProject->find('topics', [
+                Query::equal('$id', $topicIds),
+                Query::limit($topicIds)
+            ]);
             foreach ($topics as $topic) {
-                $targets = \array_filter($topic->getAttribute('targets'), fn(Document $target) => $target->getAttribute('providerType') === $message->getAttribute('providerType'));
+                $targets = \array_filter($topic->getAttribute('targets'), fn(Document $target) =>
+                    $target->getAttribute('providerType') === $message->getAttribute('providerType'));
                 $recipients = \array_merge($recipients, $targets);
             }
         }
 
-        if (\count($usersId) > 0) {
-            $users = $dbForProject->find('users', [Query::equal('$id', $usersId)]);
+        if (\count($userIds) > 0) {
+            $users = $dbForProject->find('users', [
+                Query::equal('$id', $userIds),
+                Query::limit($userIds)
+            ]);
             foreach ($users as $user) {
-                $targets = \array_filter($user->getAttribute('targets'), fn(Document $target) => $target->getAttribute('providerType') === $message->getAttribute('providerType'));
+                $targets = \array_filter($user->getAttribute('targets'), fn(Document $target) =>
+                    $target->getAttribute('providerType') === $message->getAttribute('providerType'));
                 $recipients = \array_merge($recipients, $targets);
             }
         }
 
-        if (\count($targetsId) > 0) {
-            $targets = $dbForProject->find('targets', [Query::equal('$id', $targetsId)]);
+        if (\count($targetIds) > 0) {
+            $targets = $dbForProject->find('targets', [
+                Query::equal('$id', $targetIds),
+                Query::limit($targetIds)
+            ]);
             $recipients = \array_merge($recipients, $targets);
         }
 
-        $primaryProvider = $dbForProject->findOne('providers', [
+        if (empty($recipients)) {
+            Console::error('No valid recipients found.');
+            return;
+        }
+
+        $fallback = $dbForProject->findOne('providers', [
             Query::equal('enabled', [true]),
             Query::equal('type', [$recipients[0]->getAttribute('providerType')]),
         ]);
 
+        if ($fallback === false || $fallback->isEmpty()) {
+            Console::error('No fallback provider found.');
+            return;
```

Comment:
If `processMessage()` returns like this, wouldn't the message status be stuck in `processing`?

Reply:
Fixed